File: frameworks/dotnet/carter/UnitTests/ErrorsTests.cs
namespace UnitTests;

/// <summary>
/// errors: every refusal is ASP.NET Core's own, written as ProblemDetails because the
/// Implementation asks for ProblemDetails. Carter adds nothing to any of them.
/// </summary>
public sealed class ErrorsTests(CarterApp app) : IClassFixture<CarterApp>
{
    // rb:test errors.unmatched
    [Fact]
    [Trait("corpus", "errors.unmatched")]
    public async Task A_path_no_route_matches_is_the_routers_404()
    {
        using HttpResponseMessage response = await app.CreateClient().GetAsync("/errors/unmatched");

        await Problem(HttpStatusCode.NotFound, response);
    }

    // rb:test errors.not_found
    [Fact]
    [Trait("corpus", "errors.not_found")]
    public async Task An_id_with_no_row_is_the_handlers_404()
    {
        using HttpResponseMessage response = await app.CreateClient().GetAsync("/items/999999");

        await Problem(HttpStatusCode.NotFound, response);
    }

    // rb:test errors.wrong_method
    [Fact]
    [Trait("corpus", "errors.wrong_method")]
    public async Task A_method_the_path_has_no_route_for_is_405()
    {
        using HttpResponseMessage response = await app.CreateClient().PostAsync("/items/17", null);

        await Problem(HttpStatusCode.MethodNotAllowed, response);
    }

    // rb:test errors.malformed
    [Fact]
    [Trait("corpus", "errors.malformed")]
    public async Task A_body_that_is_not_json_is_the_binders_400()
    {
        using HttpResponseMessage response = await app.CreateClient().PostAsync("/body/validate/small", new StringContent("""{"customerId": 1, "lines": [""", null, "application/json"));

        await Problem(HttpStatusCode.BadRequest, response);
    }

    private static async Task Problem(HttpStatusCode status, HttpResponseMessage response)
    {
        Assert.Equal(status, response.StatusCode);
        Assert.Equal("application/problem+json", response.Content.Headers.ContentType?.MediaType);
        Assert.Equal((int)status, (int?)(await Answer.Json(response))["status"]);
    }
}

File: frameworks/java/helidon-se/UnitTests/src/test/java/unittests/ErrorsTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

/** errors: every refusal is Helidon's own, written as text/plain by its default error handling. */
class ErrorsTests extends HelidonApp {

    // rb:test errors.unmatched
    @Test
    @Tag("errors.unmatched")
    void aPathNoRouteMatchesIsTheRouters404() throws Exception {
        refused(404, "Endpoint not found", get("/errors/unmatched"));
    }

    // rb:test errors.not_found
    @Test
    @Tag("errors.not_found")
    void anIdWithNoRowIsTheHandlersNotFoundException() throws Exception {
        refused(404, "No item has id 999999", get("/items/999999"));
    }

    // rb:test errors.wrong_method
    @Test
    @Tag("errors.wrong_method")
    void aMethodThePathHasNoRouteForIsTheSame404() throws Exception {
        refused(404, "Endpoint not found", send("POST", "/items/17", "application/json", new byte[0]));
    }

    // rb:test errors.malformed
    @Test
    @Tag("errors.malformed")
    void aBodyThatIsNotJsonIsHelidonJsonBindings400AndTheConnectionCloses() throws Exception {
        byte[] body = "{\"customerId\": 1, \"lines\": [".getBytes(StandardCharsets.UTF_8);

        HttpResponse<byte[]> response = send("POST", "/body/validate/small", "application/json", body);

        refused(400, "Failed to deserialize JSON request entity", response);
    }

    private static void refused(int status, String text, HttpResponse<byte[]> response) {
        assertEquals(status, response.statusCode());
        assertEquals("text/plain", Answer.header(response, "content-type"));
        assertEquals(text, Answer.text(response));
    }
}
